### DISCUSSION TITLE
(mini.statusline): Zero padding for the section locations.

### DISCUSSION THREAD

**Discussion by wroyca** [nvim-mini/mini.nvim#1103] [Show and tell]
Here's a little adjustment to zero-pad the section locations on the statusline. The rationale here is to avoid the component resizing continuously when moving from 10-* to 1-9 (e.g., line beginning); I also added some (optional) additional whitespace padding for the separator, making it less awkward in placement and more in line with the mode indicator visual; figured I'd share it as a snippet for anyone that'd wish to have that:

```lua
local section_location = function(args)
  -- Use virtual column number to allow update when past last column
  if MiniStatusline.is_truncated(args.trunc_width) then
    return ' ' .. '%02{max([1, line(".") - 1])}' .. ' │ %02v '
  end

  -- Use `virtcol()` to correctly handle multi-byte characters
  return ' ' .. '%02{max([1, line(".") - 1])}' .. '|%02L' .. ' │ %02v|%02{max([1, virtcol("$") - 1])} '
end
```

### Before:
![image](https://github.com/user-attachments/assets/135e45e8-de67-477a-ba3b-c1add69e0f2d)

### After:
![image](https://github.com/user-attachments/assets/958d88f9-f8ec-4c30-9c97-c877391c19e8)

**Comment by echasnovski** [comment]
Nice little adjustment!

Couple of feedback points:
- I'd use built-in Vimscript statusline strings as possible. Hence not `string.format` in full case but something like `'%02l|%L│%02v|%-2{max([1, virtcol("$") - 1])}'`.
- Not having padding for lines is deliberate because it is not that rare to have total number of lines less than 10. In which case it introduces an unnecessary width. I feel this is not the same for number columns: it is usually either 0 or more than 10.
- I'd not recommend overriding `MiniStatusline.section_location()` directly but instead add won `local location = function() ...  end` and override the whole `MiniStatusline.content.active`.

**Reply by wroyca** [reply]
> I'd use built-in Vimscript statusline strings as possible. Hence not string.format in full case but something like '%02l|%L│%02v|%-2{max([1, virtcol("$") - 1])}'.

Thanks! Attempted this at first, but wasn't comfortable with Vimscript enough, so I opted to use Lua string format. That said, I will update to Vimscript. Out of curiosity, is there a rationale here for Vimscript vs. Lua? Is it performance? If so, is there an appropriate benchmark and/or metric I can look at?

> Not having padding for lines is deliberate because it is not that rare to have total number of lines less than 10. In which case it introduces an unnecessary width

I use Neovim exclusively for programming, so my experience may influence my view here, but from day-to-day work I can say that it's actually very rare to work in files that are less than 10 lines; even tame configuration (i.e., editorconfig) typically passes the 10 lines mark easily (e.g., with comments stripped):

![image](https://github.com/user-attachments/assets/d8b5215d-f0e7-4eae-b70c-7431f690e2a7)

But again, that is, of course, from my use case, which may or may not be completely different from the usual usage. :)

> I'd not recommend overriding MiniStatusline.section_location() directly but instead add won local location = function() ...  end and override the whole MiniStatusline.content.active.

Agree; it's the approach I personally took here, but I'd recommend that whoever uses this snippet adapts it to their needs accordingly. :)

**Reply by wroyca** [reply]
Updated to `return ' %02l|%L │ %02v|%02{max([1, virtcol("$") - 1])} '`

Thank for the suggestion!

**Reply by echasnovski** [reply]
> Out of curiosity, is there a rationale here for Vimscript vs. Lua? Is it performance? If so, is there an appropriate benchmark and/or metric I can look at?

Yeah, performance. No specific public benchmarks, though. Only that I remember doing some when writing 'mini.statusline' and concluding that using special "statusline items" (as described in `:h statusline`) was faster than some alternative solutions. 

This particular case of quite fast `string.format()` might not be super relevant here, but I'd say it is most certainly **is** better to not use `vim.fn.virtcol()` here in favor of an inline `virtcol()`. Mostly because the latter is "closer to source", so to speak.

**Reply by wroyca** [reply]
> > Out of curiosity, is there a rationale here for Vimscript vs. Lua? Is it performance? If so, is there an appropriate benchmark and/or metric I can look at?
> 
> Yeah, performance. No specific public benchmarks, though. Only that I remember doing some when writing 'mini.statusline' and concluding that using special "statusline items" (as described in `:h statusline`) was faster than some alternative solutions.
> 
> This particular case of quite fast `string.format()` might not be super relevant here, but I'd say it is most certainly **is** better to not use `vim.fn.virtcol()` here in favor of an inline `virtcol()`. Mostly because the latter is "closer to source", so to speak.

All things considered, I decided to swap back the snippet to Lua out of personal preference - Vimscript syntactic is a little too uncomfortable; there's no performance impact that I can tell of, but if anything one can adapt it back to vimscript based on your comment and old revision \:)

**Reply by echasnovski** [reply]
I'd suggest to at least move `local line, virtcol = vim.fn.line, vim.fn.virtcol` to be outside of function. Otherwise it is (needlessly) executed on every statusline redraw.

That said, if you're curious for actual numbers, there is a `vim.api.nvim_eval_statusline()` which can be benchmarked. I basically use a combination of [`stat_summary()`](https://github.com/echasnovski/mini.nvim/blob/74e6b722c91113bc70d4bf67249ed8de0642b20e/doc/mini-misc.txt#L212) and [`bench_time()`](https://github.com/echasnovski/mini.nvim/blob/74e6b722c91113bc70d4bf67249ed8de0642b20e/doc/mini-misc.txt#L73).

**Reply by wroyca** [reply]
> That said, if you're curious for actual numbers, there is a vim.api.nvim_eval_statusline() which can be benchmarked. I basically use a combination of [stat_summary()](https://github.com/echasnovski/mini.nvim/blob/74e6b722c91113bc70d4bf67249ed8de0642b20e/doc/mini-misc.txt#L212) and [bench_time()](https://github.com/echasnovski/mini.nvim/blob/74e6b722c91113bc70d4bf67249ed8de0642b20e/doc/mini-misc.txt#L73).

Thank for this! Updated back to Vimscript statusline strings.

**Reply by echasnovski** [reply]
Oh, so overhead is indeed quite noticeable?

**Reply by wroyca** [reply]
> Oh, so overhead is indeed quite noticeable?

Not visibility, at least not that I can tell of on my computer, but better benchmark is always a plus :)

**Reply by echasnovski** [reply]
I was more curious about the actual numbers of Lua vs Vimscript (if you happen to remember those).

**Reply by wroyca** [reply]
```
string.format
{                                                                                                                                                                                                      
  maximum = 2.0287e-05,                                                                                                                                                                                
  mean = 5.357354e-06,                                                                                                                                                                                 
  median = 5.2295e-06,                                                                                                                                                                                 
  minimum = 4.683e-06,                                                                                                                                                                                 
  n = 1000,                                                                                                                                                                                            
  sd = 7.144184673078e-07                                                                                                                                                                              
}    

Vimscript format
{                                                                                                                                                                                                      
  maximum = 1.0413e-05,                                                                                                                                                                                
  mean = 1.795672e-06,                                                                                                                                                                                 
  median = 1.759e-06,                                                                                                                                                                                  
  minimum = 1.709e-06,                                                                                                                                                                                 
  n = 1000,                                                                                                                                                                                            
  sd = 3.6145404887128e-07                                                                                                                                                                             
}   
```

**Reply by echasnovski** [reply]
Thanks! I'd expect something like 10x difference. But 3x is also significant.
